Refactor data node startup, add startup mechanism for new node types

moesterheld opened the pull request:
## Description
Modifies the CmdLineTool to allow configurable startup of new node types.
A `GraylogNodeConfiguration` interface configures certain whether certain common startup features should be run.
For a node incorporating the features we deem as common, a `CommonNodeConfiguration` is provided which sets these flags as default.

Data node startup has been changed to use the refactored CmdLineTool directly, setting the necessary features as required by the data node. Additionally, bootstrapping of the data node has been revamped. This does not completely decouple the startup logic from Graylog startup, but uses a common codebase instead of the previously duplicated code.

An example node is currently included in the PR to showcase how new node types can be created (`ExampleCommand`). This should be removed before merging.

/nocl

## Motivation and Context
solves #18908  
solves #18011 

## How Has This Been Tested?
run data node, Graylog, certutil, journal commands manually

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring (non-breaking change)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.

bernd commented on data-node/src/main/java/org/graylog/datanode/bootstrap/DatanodeBootstrap.java:
```diff
@@ -180,24 +94,13 @@ public void configure(Binder binder) {
         final OS os = OS.getOs();
 
-        LOG.info("Graylog {} {} starting up", commandName, version);
+        LOG.info("Graylog DataNode {} {} starting up", commandName, version);
```
```suggestion
        LOG.info("Graylog Data Node {} {} starting up", commandName, version);
```

bernd commented on data-node/src/main/java/org/graylog/datanode/commands/Datanode.java:
```diff
@@ -58,29 +57,22 @@
 
 
 @Command(name = "datanode", description = "Start the Graylog DataNode")
```
```suggestion
@Command(name = "datanode", description = "Start Graylog Data Node")
```

bernd commented on data-node/src/main/java/org/graylog/datanode/bootstrap/DatanodeCmdLineTool.java:
```diff
@@ -0,0 +1,41 @@
+
+public abstract class DatanodeCmdLineTool extends CmdLineTool<Configuration> {
+
+    @Option(name = {"-f", "--configfile"}, description = "Configuration file for Graylog DataNode", override = true)
```
```suggestion
    @Option(name = {"-f", "--configfile"}, description = "Configuration file for Graylog Data Node", override = true)
```

bernd commented on data-node/src/main/java/org/graylog/datanode/bootstrap/DatanodeCmdLineTool.java:
```diff
@@ -0,0 +1,41 @@
+    @Option(name = {"-f", "--configfile"}, description = "Configuration file for Graylog DataNode", override = true)
+    protected String configFile = "/etc/graylog/datanode/datanode.conf";
+
+    @Option(name = {"-ff", "--featureflagfile"}, description = "Configuration file for Graylog DataNode feature flags", override = true)
```
```suggestion
    @Option(name = {"-ff", "--featureflagfile"}, description = "Configuration file for Graylog Data Node feature flags", override = true)
```

bernd commented on graylog2-server/src/main/java/org/graylog2/bootstrap/CmdLineTool.java:
```diff
@@ -604,8 +604,9 @@ protected void annotateInjectorExceptions(Collection<Message> messages) {
             if (rootCause instanceof NodeIdPersistenceException) {
-                LOG.error(UI.wallString(
-                        "Unable to read or persist your NodeId file. This means your node id file (" + configuration.getNodeIdFile() + ") is not readable or writable by the current user. The following exception might give more information: " + message));
+                //TODO: Check configuration
+//                LOG.error(UI.wallString(
+//                        "Unable to read or persist your NodeId file. This means your node id file (" + configuration.getNodeIdFile() + ") is not readable or writable by the current user. The following exception might give more information: " + message));
```
We should fix this TODO item.

bernd commented on graylog2-server/src/main/java/org/graylog2/commands/ExampleCommand.java:
We should remove this file.

moesterheld replied:
Yes, I wanted to remove that in the end (see PR comment).
I have now moved it to a test. I will extend on that test to include different node configurations to make it more complete.

bernd commented on graylog2-server/src/main/java/org/graylog2/commands/ExampleCommandConfiguration.java:
We should remove this file.

bernd commented on graylog2-server/src/main/java/org/graylog2/commands/ServerCommandsProvider.java:
```diff
@@ -31,6 +31,7 @@ public void addTopLevelCommandsOrGroups(CliBuilder<CliCommand> builder) {
 
         builder.withCommand(Server.class);
         builder.withCommand(MigrateCmd.class);
+        builder.withCommand(ExampleCommand.class);
```
Since we should remove the example files, we shouldn't bind the example command here.

bernd reviewed (changes requested):
A first quick review. Will do testing and a more detailed review next.

bernd commented on graylog2-server/src/main/java/org/graylog2/CommonNodeConfiguration.java:
```diff
@@ -0,0 +1,72 @@
+    }
+
+    @Override
+    default boolean isMessageRecordingsEnabled() {
```
Let's add an `@Override` annotation to the `BaseConfiguration#isMessageRecordingsEnabled` method.

bernd commented on data-node/src/main/java/org/graylog/datanode/commands/Datanode.java:
```diff
@@ -57,30 +56,23 @@
+public class Datanode extends DatanodeBootstrap {
+    private static final Logger LOG = LoggerFactory.getLogger(Datanode.class);
 
     protected static final Configuration configuration = new Configuration();
```
The `configuration` field can be private. Does it really need to be static?

bernd commented on data-node/src/main/java/org/graylog/datanode/commands/Datanode.java:
```diff
@@ -57,30 +56,23 @@
     }
 
-    public Server(String commandName) {
+    public Datanode(String commandName) {
```
This constructor is unused.

bernd commented on data-node/src/main/java/org/graylog/datanode/bootstrap/DatanodeBootstrap.java:
```diff
@@ -180,24 +94,13 @@ public void configure(Binder binder) {
         final OS os = OS.getOs();
 
-        LOG.info("Graylog {} {} starting up", commandName, version);
+        LOG.info("Graylog Data Node {} {} starting up", commandName, version);
```
The log message shows redundant information:

```
[DatanodeBootstrap] Graylog Data Node datanode 6.2.0-SNAPSHOT starting up
```

How about explicitly describing the command? Or we leave the command out.

```suggestion
        LOG.info("Graylog Data Node {} starting up (command: {})", version, commandName);
```

bernd commented on data-node/src/main/java/org/graylog/datanode/bootstrap/DatanodeBootstrap.java:
```diff
@@ -34,89 +32,35 @@
-        }
-        // This needs to run before the first SSLContext is instantiated,
-        // because it sets up the default SSLAlgorithmConstraints
-        applySecuritySettings(tlsProtocolsConfiguration);
```
Do we not need the security settings for Data Node?

bernd replied:
Ping.

moesterheld replied:
You're right. I moved that to the general `beforeStart` now since setting the values doesn't hurt for all kinds of nodes.

bernd commented on graylog2-server/src/main/java/org/graylog2/CommonNodeConfiguration.java:
```diff
@@ -0,0 +1,72 @@
+    @Override
+    default String getEnvironmentVariablePrefix() {
+        return "GRAYLOG_";
+    };
```
```suggestion
    }
```

bernd commented on graylog2-server/src/main/java/org/graylog2/CommonNodeConfiguration.java:
```diff
@@ -0,0 +1,72 @@
+    @Override
+    default String getSystemPropertyPrefix() {
+        return "graylog.";
+    };
```
```suggestion
    }
```

bernd commented on graylog2-server/src/main/java/org/graylog2/bindings/MongoDbConnectionModule.java:
```diff
@@ -0,0 +1,45 @@
+ */
+package org.graylog2.bindings;
+
+import com.google.inject.AbstractModule;
```
Unused import.

bernd commented on graylog2-server/src/main/java/org/graylog2/bindings/MongoDbConnectionModule.java:
```diff
@@ -0,0 +1,45 @@
+import org.graylog2.database.MongoConnection;
+import org.graylog2.plugin.inject.Graylog2Module;
+
+import java.util.List;
```
Unused import.

bernd commented on graylog2-server/src/main/java/org/graylog2/bindings/GraylogNodeModule.java:
```diff
@@ -0,0 +1,99 @@
+/**
+ * <p>Guice module that contains all necessary bindings to start a basic node in a Graylog cluster.</p>
+ *
+ * <ul>
```
Let's either update the list of bindings to match the module or remove the list.

bernd replied:
Ping.

moesterheld replied:
i removed the list here and added documentation to the config toggles in `GraylogNodeConfiguration`

bernd replied:
Thanks!

bernd commented on graylog2-server/src/main/java/org/graylog2/bindings/GraylogNodeModule.java:
```diff
@@ -0,0 +1,99 @@
+        install(new AuditBindings());
+
+        if (configuration.withEventBus()) {
+            bind(EventBus.class).toProvider(EventBusProvider.class).in(Scopes.SINGLETON);
```
We still have the `org.graylog.datanode.bindings.ServerBindings` file that also binds the event bus. (there are more redundancies, also in the server's `ServerBindings` file) Can we get rid of the `ServerBindings` file? If not we should rename it to `DataNodeBindings`.

bernd commented on graylog2-server/src/main/java/org/graylog2/bindings/GraylogNodeModule.java:
Please check if we have bindings from this file in other files. A quick check showed several redundancies in other modules, which we can probably remove.

bernd commented on graylog2-server/src/main/java/org/graylog2/bindings/MongoDbConnectionModule.java:
It looks like there are redundancies with the `MongoDBModule`.

bernd commented on graylog2-server/src/main/java/org/graylog2/commands/AbstractNodeCommand.java:
```diff
@@ -0,0 +1,68 @@
+/**
+ * Inherit from this command to create new standalone node types.
+ */
+public abstract class AbstractNodeCommand extends CmdLineTool<GraylogNodeConfiguration> {
```
Why do the Server and Data Node nodes not extend this class?

bernd reviewed (changes requested):
Thanks! A few comments.

moesterheld commented:
@bernd I addressed your comments. 
I also cleaned up the redundant bindings and made all commands, including the cli commands use the AbstractNodeCommand. The only redundant bindings are now for preflight. In the future, we could think about making preflight bindings configurable from a GraylogNodePreflightModule and/or getPreflightCommandBindings methods in the AbstractNodeCommand.

bernd reviewed (changes requested):
I left two pings in comments that haven't been answered. (or I didn't see it)

bernd reviewed (approved):
Thank you! :pray:

thll commented on graylog2-server/src/main/java/org/graylog2/bindings/GraylogNodeModule.java:
```diff
@@ -0,0 +1,95 @@
+        install(new ServerStatusBindings(configuration.withCapabilities()));
+
+        bind(EncryptedValueService.class).asEagerSingleton();
+        bind(InputConfigurationBeanDeserializerModifier.class).toInstance(InputConfigurationBeanDeserializerModifier.withoutConfig());
```
@moesterheld: For the server, this binding breaks the saving of inputs that contain encrypted values. You can reproduce it by creating an `AWS CloudTrail` where you only set the title (and node) and leave everything else empty. You can save it, but afterwards the inputs page is broken and you have to remove the input manually from MongoDB.

We should either go back to a just-in-time binding (for the server) or do a regular singleton scope binding. I didn't test the latter but it should be fine.

moesterheld replied:
Thanks @thll , I will fix this and try to add an automated test that covers this somehow.

thll replied:
Thank you! Yeah, having some kind of integration test for this would be great as the cause is very subtle and hard to find.
